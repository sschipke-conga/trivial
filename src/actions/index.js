export const setTeams = (teams) => ({
  type: 'SET_TEAMS',
  teams
})

export const setQuestions = (questions) => ({
  type: 'SET_QUESTIONS',
  questions
})