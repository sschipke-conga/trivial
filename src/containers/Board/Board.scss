.board {
  background: url(../../images/question.jpg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  color: white;
  display: flex;
  justify-content: center;
  height: 100vh;
  padding: 20px;
  width: 100vw;
}
.question-div {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: row;
  flex-wrap: wrap;
  width: 50%;
  .question-headers {
    display: flex;
    justify-content: center;
    margin-bottom: 15px;
    min-width: 80vw;
    .category {
      width: 40%;
    }
  }
  h3 {
    width: 20%;
  }
  .question {
    text-align: center;
    width: 100%;
  }
  .question-header-div {
    border: 2px solid white;
    background: rgba(0, 0, 0, 0.85);
    justify-content: center;
    align-items: center;
    padding: 10px;
    }
}

.answers-div {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 20px 0 20px 0;
  .current-team {
    font-size: 1.5em;
  }
}