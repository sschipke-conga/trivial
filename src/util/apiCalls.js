export const fetchQuestions = async (amount, category, difficulty) => {
  let url = `https://opentdb.com/api.php?amount=${amount}&category=${category}&difficulty=${difficulty}&type=multiple`
  const res = await fetch(url);
  if (!res.ok) {
    throw Error('Failed to fetch')
  }
  const data = await res.json()
  return data.results
}
