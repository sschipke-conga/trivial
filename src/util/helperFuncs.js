export const shuffle = answers => {
  var m = answers.length, t, i;
  while (m) {
    i = Math.floor(Math.random() * m--);
    t = answers[m];
    answers[m] = answers[i];
    answers[i] = t;
  }
  return answers;
}

export const determineCurrentTeam = (turnCount) => {
  if (turnCount % 2 === 0) {
    return 1
  } 
  return 0
}

export const determineWinner = teams => {
  if(teams[0].score === teams[1].score) {
    return false
  }
  return teams.sort((teamOne, teamTwo) => {
    return teamTwo.score-teamOne.score
  })[0]
}

export const capDifficulty = difficulty => {
  return difficulty[0].toUpperCase() + difficulty.slice(1)
}

export const cleanEncoding = (string) => {
  let cleanPhrase = string.replace(/&#039;/g, "'")
  .replace(/&quot;/g, '"')
  .replace(/&eacute/g, 'Ã©');
  return cleanPhrase
}

export const cleanQuestions = (questions) => {
    return questions.map(question => {
    question.question = cleanEncoding(question.question)
    question.correct_answer = cleanEncoding(question.correct_answer)
    question.incorrect_answers = question.incorrect_answers.map(answer => cleanEncoding(answer))
    return question
    })
}