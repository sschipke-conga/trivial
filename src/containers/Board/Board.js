import React, {Component} from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {determineCurrentTeam, capDifficulty} from '../../util/helperFuncs'
import {setCurrentQuestion, setQuestions, increaseTurnCount, setCurrentTeam, updateScore, setHaveWinner} from '../../actions/index';
import Answers from '../../components/Answers/Answers';
import './Board.scss';
export class Board extends Component {
  constructor() {
    super()
  }

  componentDidMount() {
    const {increaseTurnCount, turnCount, teams, setCurrentTeam} = this.props
    increaseTurnCount();
    this.updateCurrentQuestion()
    let newTurnCount = turnCount + 1
    const currentTeam = teams[determineCurrentTeam(newTurnCount)];
    setCurrentTeam(currentTeam)
  }

  updateCurrentQuestion = () => {
    const {questions, currentQuestion, setCurrentQuestion, setQuestions} = this.props;
    const selectedCurrentQuestion = questions.pop();
    setQuestions(questions);
    setCurrentQuestion(selectedCurrentQuestion);
  }

  submitAnswer = (e) => {
    let {currentQuestion, increaseTurnCount, currentTeam, setCurrentTeam, teams, turnCount, updateScore, questions, setHaveWinner} = this.props
    if (e.target.name === currentQuestion.correct_answer) {
      currentTeam.score++
      updateScore()
    }
    if (!questions.length) {
      setHaveWinner()
      return null;
    }
    increaseTurnCount()
    this.updateCurrentQuestion()
    let newTurnCount = turnCount + 1
    currentTeam = teams[determineCurrentTeam(newTurnCount)];
    setCurrentTeam(currentTeam)
  }

  render() {
    const {teams, currentQuestion, currentTeam, questions} = this.props
    return (
      <div className="board">
        <div className="board-team-one">
          <h1 className="team-one-name">{teams[0].name}</h1>
          <h3 className="score">Score: {teams[0].score}</h3>
        </div>
        <div className="question-div">
          <div className="question-header-div">
            <div className="question-headers">
              <h3 className="category">{currentQuestion.category}</h3>
              <h3 className="difficulty">{capDifficulty(currentQuestion.difficulty)}</h3>
            </div>
            <h2 className="question">{currentQuestion.question}</h2>
            <div className="answers"><Answers submitAnswer={this.submitAnswer} answers={[currentQuestion.correct_answer, ...currentQuestion.incorrect_answers]}/>
            <h4>{currentTeam.name}, your turn!</h4>
            {questions.length !== 0 && <h5>{questions.length+1} questions left</h5>}
            {questions.length === 0 && <h5>Last question!</h5>}
            </div>
          </div>
        </div>
        <div className="board-team-two">
          <h1 className="team-one-name">{teams[1].name}</h1>
          <h3 className="score">Score: {teams[1].score}</h3>
        </div>
      </div>
    )
  } 
}

export const mapStateToProps = state => ({
  questions: state.questions,
  teams: state.teams,
  currentQuestion: state.currentQuestion,
  turnCount: state.turnCount,
  currentTeam: state.currentTeam
})

export const mapDispatchToProps = dispatch => bindActionCreators({setQuestions, setCurrentQuestion, increaseTurnCount, setCurrentTeam, updateScore, setHaveWinner}, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(Board)