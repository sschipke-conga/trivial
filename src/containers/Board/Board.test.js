import React from 'react';
import { Board, mapDispatchToProps, mapStateToProps } from './Board';
import { shallow } from 'enzyme';
import * as actions from '../../actions';
import * as helpers from '../../util/helperFuncs';

jest.mock('../../util/helperFuncs')

jest.mock('../../actions')
let mockQuestions = [{
  "category": "Sports",
  "type": "multiple",
  "difficulty": "medium",
  "question": "At which bridge does the annual Oxford and Cambridge boat race start?",
  "correct_answer": "Putney",
  "incorrect_answers": [
    "Hammersmith",
    "Vauxhall ",
    "Battersea"
  ]
},
{
  "category": "Entertainment: Video Games",
  "type": "multiple",
  "difficulty": "easy",
  "question": "Who is Sonic&#039;s sidekick?",
  "correct_answer": "Tails",
  "incorrect_answers": [
    "Shadow",
    "Amy",
    "Knuckles"
  ]
},
{
  "category": "Politics",
  "type": "boolean",
  "difficulty": "easy",
  "question": "There was a satirical candidate named &quot;Deez Nuts&quot; running in the 2016 US presidential elections.",
  "correct_answer": "True",
  "incorrect_answers": [
    "False"
  ]
}]
const mockCurrentQuestion = {
  "category": "Politics",
  "type": "boolean",
  "difficulty": "easy",
  "question": "There was a satirical candidate named &quot;Deez Nuts&quot; running in the 2016 US presidential elections.",
  "correct_answer": "True",
  "incorrect_answers": [
    "False"
  ]
}

const mockTeams = [{
  name: 'Lez Quizzerable',
  score: 0
}, {
  name: "Let's Get Quizzacle",
  score: 0
}]

actions.setTeams = jest.fn().mockImplementation(() => {
  return mockTeams
})

actions.setHaveQuestions = jest.fn()

actions.setQuestions = jest.fn().mockImplementation(() => mockQuestions)

actions.setCurrentQuestion = jest.fn().mockImplementation(() => mockCurrentQuestion)

helpers.determineCurrentTeam.mockImplementation(() => 0)

describe('Board', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Board setQuestions={actions.setQuestions} setCurrentQuestion={actions.setCurrentQuestion}
      increaseTurnCount={actions.increaseTurnCount}
      setCurrentTeam={actions.setCurrentTeam}
      updateScore={actions.updateScore}
      setHaveWinner={actions.setHaveWinner}
      questions={mockQuestions}
      teams={mockTeams}
      currentQuestion={mockCurrentQuestion}
      turnCount={0}
      currentTeam={mockTeams[0]}
      />
    )
  })
  it('should match the initial snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  describe('Board componentDidMount',() => {
    it('should call increaseTurnCount after mounting', () => {
      expect(actions.increaseTurnCount).toHaveBeenCalled()
    })
    it('should call updateCurrentQuestion after mounting', () => {
      wrapper.instance().updateCurrentQuestion = jest.fn()
      wrapper.instance().forceUpdate()
      wrapper.instance().componentDidMount()
      expect(wrapper.instance().updateCurrentQuestion).toHaveBeenCalled()
    })
    it('should call determineCurrentTeam after mounting', () => {
      expect(helpers.determineCurrentTeam).toHaveBeenCalled()
    })
    it('should call setCurrentTeam after mounting', () => {
      expect(actions.setCurrentTeam).toHaveBeenCalledWith(mockTeams[0])
    })
  })
})