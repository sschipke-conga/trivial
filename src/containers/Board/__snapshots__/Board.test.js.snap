// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should match the initial snapshot 1`] = `
<div
  className="board"
>
  <div
    className="board-team-one"
  >
    <h1
      className="team-one-name"
    >
      Lez Quizzerable
    </h1>
    <h3
      className="score"
    >
      Score: 
      0
    </h3>
  </div>
  <div
    className="question-div"
  >
    <div
      className="question-header-div"
    >
      <div
        className="question-headers"
      >
        <h3
          className="category"
        >
          Politics
        </h3>
        <h3
          className="difficulty"
        />
      </div>
      <h2
        className="question"
      >
        There was a satirical candidate named &quot;Deez Nuts&quot; running in the 2016 US presidential elections.
      </h2>
      <div
        className="answers-div"
      >
        <Answers
          answers={
            Array [
              "True",
              "False",
            ]
          }
          submitAnswer={[Function]}
        />
        <h4>
          <span
            className="current-team"
          >
            Lez Quizzerable
          </span>
          , your turn!
        </h4>
        <h5>
          4
           questions left
        </h5>
      </div>
    </div>
  </div>
  <div
    className="board-team-two"
  >
    <h1
      className="team-one-name"
    >
      Let's Get Quizzacle
    </h1>
    <h3
      className="score"
    >
      Score: 
      0
    </h3>
  </div>
</div>
`;
